crates/oxc_js_regex/src/ast.rs

```diff
@@ -0,0 +1,387 @@
+}
+
+#[derive(Debug)]
+pub struct CharacterUnicodePropertyCharacterSet {
+    pub span: Span,
+    pub key: Atom,
+    pub value: Option<Atom>,
+    pub negate: bool,
+}
+
+/// StringsUnicodePropertyCharacterSet is Unicode property escape with property of strings.
+#[derive(Debug)]
+pub struct StringsUnicodePropertyCharacterSet {
+    pub span: Span,
+    pub key: Atom,
+}
+
+/// The expression character class.
+/// E.g. `[a--b]`, `[a&&b]`,`[^a--b]`, `[^a&&b]`
+#[derive(Debug)]
+pub struct ExpressionCharacterClass<'a> {
+    pub span: Span,
+    pub negate: bool,
+    pub expression: ExpressionCharacterClassExpr<'a>,
+}
+
+#[derive(Debug)]
+pub enum ExpressionCharacterClassExpr<'a> {
+    ClassIntersection(Box<'a, ClassIntersection<'a>>),
+    ClassSubtraction(Box<'a, ClassSubtraction<'a>>),
+}
+
+#[derive(Debug)]
+pub enum ClassSetOperand<'a> {
+    Character(Box<'a, Character>),
+    ClassStringDisjunction(Box<'a, ClassStringDisjunction<'a>>),
+    EscapeCharacterSet(Box<'a, EscapeCharacterSet>),
+    ExpressionCharacterClass(Box<'a, ExpressionCharacterClass<'a>>),
+    UnicodePropertyCharacterSet(Box<'a, UnicodePropertyCharacterSet<'a>>),
+    UnicodeSetsCharacterClass(Box<'a, UnicodeSetsCharacterClass<'a>>),
+}
+
+/// The character class intersection.
+/// E.g. `a&&b`
+#[derive(Debug)]
+pub struct ClassIntersection<'a> {
+    pub span: Span,
+    pub left: ClassIntersectionLeft<'a>,
+    pub right: ClassSetOperand<'a>,
+}
+
+#[derive(Debug)]
+pub enum ClassIntersectionLeft<'a> {
+    ClassIntersection(Box<'a, ClassIntersection<'a>>),
+    ClassSetOperand(Box<'a, ClassSetOperand<'a>>),
+}
+
+/// The character class subtraction.
+/// E.g. `a--b`
+#[derive(Debug)]
+pub struct ClassSubtraction<'a> {
+    pub span: Span,
+    pub left: ClassSubtractionLeft<'a>,
+    pub right: ClassSetOperand<'a>,
+}
+
+#[derive(Debug)]
+pub enum ClassSubtractionLeft<'a> {
+    ClassSetOperand(Box<'a, ClassSetOperand<'a>>),
+    ClassSubtraction(Box<'a, ClassSubtraction<'a>>),
+}
+
+/// The character class string disjunction.
+/// E.g. `\q{a|b}`
+#[derive(Debug)]
+pub struct ClassStringDisjunction<'a> {
+    pub span: Span,
+    pub alternatives: Vec<'a, StringAlternative<'a>>,
+}
+
+/// StringAlternative is only used for `\q{alt}`({@link ClassStringDisjunction}).
+#[derive(Debug)]
+pub struct StringAlternative<'a> {
+    pub span: Span,
+    pub elements: Vec<'a, Character>,
+}
+
+/// This includes escape sequences which mean a character.
+/// E.g. `a`, `あ`, `✿`, `\x65`, `\u0065`, `\u{65}`, `\/`
+#[derive(Debug)]
+pub struct Character {
+    pub span: Span,
+    pub value: u16, // UTF-16 code point
+}
+
+#[derive(Debug)]
+pub enum BackreferenceRef {
+    Number(i32),
+    Atom(Atom),
+}
+
+/// The backreference.
+/// E.g. `\1`, `\k<name>`
+#[derive(Debug)]
+pub struct Backreference<'a> {
+    pub span: Span,
+    pub reference: BackreferenceRef,
+    pub resolved: CapturingGroup<'a>,
+}
+
+/// The flags.
+#[derive(Debug)]
+pub struct Flags {
```

Comment:
`bitflags` is a better choice here, since it could save more space, you could refer https://github.com/oxc-project/oxc/blob/c934d93713b5b5fd577d704a9e5063518e546719/crates/oxc_semantic/src/node.rs#L13-L21 for more details.

Reply:
Thank you! I didn't know that (I should have looked at other implementations a bit more).

---

While I was exploring how to use bit flags, I accidentally stumbled upon a similar implementation for flags in the JS AST. 

So, it's okay to disregard this and go ahead with a unique implementation in js_regex, right? 
(Seems like the only option?)

The crate I've created is intended to provide a more detailed Regex AST and parser.

https://github.com/oxc-project/oxc/blob/c934d93713b5b5fd577d704a9e5063518e546719/crates/oxc_ast/src/ast/literal.rs#L142-L155

Reply:
@Boshen

Reply:
We can change this later.